core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -355,10 +355,8 @@ class KafkaApis(val requestChannel: RequestChannel,
             offsetMetadata = OffsetMetadata(partitionData.offset, metadata),
             commitTimestamp = currentTimestamp,
             expireTimestamp = {
-              if (header.apiVersion >= 3)
+              if (header.apiVersion >= 2 || partitionData.timestamp == OffsetCommitRequest.DEFAULT_TIMESTAMP)
```

Comment:
I am not sure about this check. I think we should use `None` for every case except when the version is v2 and the retention time is set to use the default. In fact, can't we just do this?
```
val retentionTime = offsetCommitRequest.retentionTime
if (retentionTime == DEFAULT_RETENTION_TIME)
  Some(currentTimestamp + retentionTime)
else
  None
```
Only v2 requests will have anything other than `DEFAULT_RETENTION_TIME`. For everything else, we'll use the new expiration semantics.

Reply:
For v2/v3/v4 the retention time could be something other than the default.
For v1 the commit time could be something other than the default, in which case we calculate an expiration timestamp.
The expiration time would be `None` if retention time is the default.
```
            expireTimestamp = offsetCommitRequest.retentionTime match {
              case OffsetCommitRequest.DEFAULT_RETENTION_TIME => None
              case retentionTime =>
                partitionData.timestamp match {
                  case OffsetCommitRequest.DEFAULT_TIMESTAMP => None
                  case commitTimestamp => Some(commitTimestamp + retentionTime)
                }
            }
```
I used the above for expiration timestamp. Let me know if that works.